src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java

```diff
@@ -144,18 +144,25 @@ private URI addApiCallQueryParameters(URI url) throws NotLoggedInException {
         return url;
     }
 
-    public Callable<Exercise> getAdaptiveExercise() 
-        throws OAuthSystemException, OAuthProblemException, NotLoggedInException {
-        return wrapWithNotLoggedInException(new Callable<Exercise>() {
+    public URI getSkillifierUrl(String addition) {
+        if (!addition.isEmpty()) {
+            return URI.create("http://tmc-adapt.testmycode.io/Example/default/" + addition);
+        }
+        return URI.create("http://tmc-adapt.testmycode.io/Example/default/");
+    }
+
+    public Callable<AdaptiveExercise> getAdaptiveExercise()
+            throws OAuthSystemException, OAuthProblemException, NotLoggedInException {
+        return wrapWithNotLoggedInException(new Callable<AdaptiveExercise>() {
             @Override
-            public Exercise call() throws Exception {
+            public AdaptiveExercise call() throws Exception {
                 try {
                     Callable<String> download = new HttpTasks()
-                                        .getForText(URI.create(SKILLIFIER_URL));
+                                        .getForText(getSkillifierUrl("next.json?username=" + oauth.getToken()));
                     String json = download.call();
                     return adaptiveExerciseParser.parseFromJson(json);
                 } catch (Exception ex) {
-                    LOG.log(Level.WARNING, "Downloading and parsing adaptive exercise URL failed with exception: " + ex.getMessage(), ex);
+                    LOG.log(Level.WARNING, "Downloading and parsing adaptive exercise URL failed.");
```

Comment:
Log the full exception (that is, its stack trace) unless there's some specific reason not to.